File: python_version/endmove.py
from random import choice, uniform
from math import exp

class EndMove(object):
    def __init__(self, space, beta):
        self.space = space
        self.movetype = "EndMove"
        self.beta = beta
    def move(self, polyid, polytyp, endtyp):
        """
        The end move
        """
        if polytyp == 'sim':
            poly = self.space.Sims[polyid-1]
            sitevalue = 1 # code for free sim
        elif polytyp == 'sumo':
            poly = self.space.Sumos[polyid-1]
            sitevalue = -1 # code for free sumo
        else:
            raise Exception("polymer type undefined")
            
        if endtyp == 'head':
            endloc, nextloc = 0, 1
        elif endtyp == 'tail':
            endloc, nextloc = len(poly.locs)-1, len(poly.locs)-2
        else:
            raise Exception("end type undefined")
            
        assert(self.space.space[poly.locs[endloc][0]][poly.locs[endloc][1]] != 0)
            
        newpoint = choice(list(self.space.neighbor(poly.locs[nextloc]) - set([poly.locs[endloc]])))
        
        x, y = newpoint
        if self.space.space[x][y] != 0: # site occupied, not possible to move
            return

        xold, yold = poly.locs[endloc] 
        
        nbr_bond_inc = 0
        
        if abs(self.space.space[xold][yold]) != 1: # (xold, yold) is in a bond
            nbr_bond_inc -= 1 # will lose a bond
            
        bond_choice = [bpoint for bpoint in self.space.neighbor(newpoint) if self.space.can_build_bond(newpoint, bpoint)]
        if bond_choice: # (x, y) can build a bond
            nbr_bond_inc += 1 # will create a bond
        
        if uniform(0,1) < self.weight(nbr_bond_inc, self.beta): # 
            # removes the monomer on (xold, yold)
            # handles everything except in its polymer.locs and rspace
            self.space.safe_remove(xold, yold) 
            
            # make changes on polymer.locs
            poly.locs[endloc] = newpoint
            
            # create a monomer on (x, y)
            # handles everything except in its polymer.locs and rspace
            self.space.safe_create(x, y, sitevalue)
            
            # handle the reverse checking space
            assert(self.space.rspace[x][y][0] == 0 and self.space.rspace[xold][yold][0] != 0)
            self.space.rspace[x][y][0] = self.space.rspace[xold][yold][0]
            self.space.rspace[x][y][1] = self.space.rspace[xold][yold][1]
            self.space.rspace[xold][yold][0], self.space.rspace[xold][yold][1] = 0, 0
            
            # try building new bonds
            
            if bond_choice:
                (xb, yb) = choice(bond_choice)
                self.space.create_bond((x, y), (xb, yb))
            
            
    def weight(self, nbr_bond_inc, beta):
        """
        The weight function for endmove
        """
        return exp(nbr_bond_inc*beta)

File: Julia_version/EndMove.jl
module EndMoveModule
export EndMove, move


include("Space.jl")
using SpaceModule

function weight(nbr_bond_inc, beta)
    exp(nbr_bond_inc*beta)
end

type EndMove
    beta :: Float64
    movetype :: ASCIIString
    EndMove() = new(0, "EndMove")
end

function avail_filter(space::Space, point::Tuple{Int64,Int64})
    space.space[point[1], point[2]] == 0
end

function get_possible_moves(move::EndMove, space::Space, poly::Polymer)
    # This functions is almost identical in different spaces except for the avail_filter
    endloc_head, nextloc_head = 1, 2
    endloc_tail, nextloc_tail = length(poly.locs), length(poly.locs)-1
    
    possible_moves_head = neighbor(space, poly.locs[nextloc_head])
    possible_moves_tail = neighbor(space, poly.locs[nextloc_tail])

    # this automatically chooses the filter for that geometry
    avail_f(point) = avail_filter(space, point) 
    
    moves_head = filter(avail_f, possible_moves_head)
    moves_tail = filter(avail_f, possible_moves_tail)

    append!([(endloc_head, point) for point in moves_head] , 
    [(endloc_tail, point) for point in moves_tail] )
    
end

function update_reverse_checking_space(move::EndMove, space::Space, oldpoint, newpoint)
    # handle the reverse checking space
    xnew, ynew = newpoint
    xold, yold = oldpoint
    assert(space.rspace[xnew, ynew, 1] == 0 && space.rspace[xold, yold, 1] != 0)
    space.rspace[xnew, ynew, 1] = space.rspace[xold, yold, 1]
    space.rspace[xnew, ynew, 2] = space.rspace[xold, yold, 2]
    space.rspace[xold, yold, 1], space.rspace[xold, yold, 2] = 0, 0
end

function move(move::EndMove, space::Space, polyid::Int, polytyp::ASCIIString)
    if polytyp == "sim"
        poly = space.Sims[polyid]
        sitevalue = 1 # code for free sim
    elseif polytyp == "sumo"
        poly = space.Sumos[polyid]
        sitevalue = -1 # code for free sumo
    else
        error("polymer type undefined")
    end
    
    possible_moves = get_possible_moves(move, space, poly)
    
    if isempty(possible_moves)
        return
    end

    (pointid, newpoint) = possible_moves[rand(1:end)]
    oldpoint = poly.locs[pointid]       
    
    nbr_bond_inc = 0
    
    if in_a_bond(space, oldpoint) # (xold, yold) is in a bond
        nbr_bond_inc -= 1 # will lose a bond
    end
    
    bond_filter(bpoint) = can_build_bond(space, newpoint, bpoint)
    bond_choice = filter(bond_filter, neighbor(space, newpoint))
    
    if !isempty(bond_choice) # (x, y) can build a bond
        nbr_bond_inc += 1 # will create a bond
    end
    
    if rand() < weight(nbr_bond_inc, move.beta) 
        # removes the monomer on oldpoint
        # handles everything except in its polymer.locs and rspace
        safe_remove(space, oldpoint) 

        # create a monomer on newpoint
        # handles everything except in its polymer.locs and rspace
        safe_create(space, newpoint, sitevalue)

        # make changes on polymer.locs
        poly.locs[pointid] = newpoint

        # handle the reverse checking space (This part needs to use specific geometry)
        update_reverse_checking_space(move, space, oldpoint, newpoint)

        # try building new bonds
        if !isempty(bond_choice)
            bpoint = choice(bond_choice)
            create_bond(space, newpoint, bpoint)
        end
    end

end

end
